# 🏥 Medical Insurance Cost Prediction

This project explores healthcare analytics using a medical insurance dataset. The goal is to understand how demographic and health factors influence medical costs and to develop models for prediction, classification, and clustering.

---

## 📌 Objectives
- Predict annual medical insurance charges (Regression)
- Classify individuals by health/risk level
- Cluster similar health and insurance profiles
- Analyze relationships between demographics, health factors, and costs

---
## 🧰 Tools & Libraries
- Python, Pandas, NumPy  
- Scikit-learn  
- Matplotlib, Seaborn  
- Jupyter Notebook

### 📦 Project Libraries - install

In [1]:
%pip install -q -U pandas numpy matplotlib seaborn scikit-learn  
# Install and upgrade essential packages:
# pandas & numpy -> data manipulation and numerical operations
# matplotlib & seaborn -> data visualization
# scikit-learn -> machine learning (regression, classification, clustering)

%pip install -q -U watermark  
# Install/upgrade watermark to display package versions and environment info

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 25.2 -> 25.3
[notice] To update, run: python.exe -m pip install --upgrade pip


Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 25.2 -> 25.3
[notice] To update, run: python.exe -m pip install --upgrade pip


### 📦 Project Libraries - import

This section lists all the Python libraries used in this project. 
Keeping them organized here helps with reproducibility and makes it easier to install dependencies.

In [2]:
# Importing the library for data manipulation in tables
import pandas as pd 

# Importing the NumPy library for mathematical operations and arrays
import numpy as np  

# Importing the Matplotlib library for generating plots
import matplotlib.pyplot as plt

# Jupyter Notebook magic command to display plots directly in the notebook
%matplotlib inline

# Importing the Seaborn library for statistical data visualization
import seaborn as sns

# scikit learn area:

# Import for regression
from sklearn.linear_model import LinearRegression  

# Import for classification
from sklearn.ensemble import RandomForestClassifier  

# Import for clustering
from sklearn.cluster import KMeans  

# Import for splitting dataset
from sklearn.model_selection import train_test_split  

# Import for evaluating models
from sklearn.metrics import mean_squared_error, accuracy_score



# Load the watermark extension
%reload_ext watermark

# Display metadata for your notebook
%watermark -a "Maykon Analysis" -d -u -v -p numpy,pandas,matplotlib,seaborn,scikit-learn

Author: Maykon Analysis

Last updated: 2025-11-24

Python implementation: CPython
Python version       : 3.13.7
IPython version      : 9.7.0

numpy       : 2.3.5
pandas      : 2.3.3
matplotlib  : 3.10.7
seaborn     : 0.13.2
scikit-learn: 1.7.2



#### Loading a dataset into your working environment (using the pandas library).

In [3]:
# load the data
data = r"C:\Users\LarTI\OneDrive\Desktop\Projects\Medical-Insurance-Cost-Prediction-\data\medical_insurance.csv"

df = pd.read_csv(data)

In [4]:
df.head()

person_id  age     sex   region urban_rural   income     education  \
0      75722   52  Female    North    Suburban  22700.0     Doctorate   
1      80185   79  Female    North       Urban  12800.0         No HS   
2      19865   68    Male    North       Rural  40700.0            HS   
3      76700   15    Male    North    Suburban  15600.0  Some College   
4      92992   53    Male  Central    Suburban  89600.0     Doctorate   

  marital_status employment_status  household_size  ...  liver_disease  \
0        Married           Retired               3  ...              0   
1        Married          Employed               3  ...              0   
2        Married           Retired               5  ...              0   
3        Married     Self-employed               5  ...              0   
4        Married     Self-employed               2  ...              0   

   arthritis mental_health proc_imaging_count  proc_surgery_count  \
0          1             0                  1                   0   
1          1             1                  0                   0   
2          0             1                  1                   0   
3          0             0                  1                   0   
4          1             0                  2                   0   

   proc_physio_count  proc_consult_count  proc_lab_count  is_high_risk  \
0                  2                   0               1             0   
1                  1                   0               1             1   
2                  2                   1               0             1   
3                  0                   1               0             0   
4                  1                   1               0             1   

   had_major_procedure  
0                    0  
1                    0  
2                    0  
3                    0  
4                    0  

[5 rows x 54 columns]

In [5]:
df.info()   # shows column types and non-null counts

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 54 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   person_id                    100000 non-null  int64  
 1   age                          100000 non-null  int64  
 2   sex                          100000 non-null  object 
 3   region                       100000 non-null  object 
 4   urban_rural                  100000 non-null  object 
 5   income                       100000 non-null  float64
 6   education                    100000 non-null  object 
 7   marital_status               100000 non-null  object 
 8   employment_status            100000 non-null  object 
 9   household_size               100000 non-null  int64  
 10  dependents                   100000 non-null  int64  
 11  bmi                          100000 non-null  float64
 12  smoker                       100000 non-null  object 
 13  

In [ ]:
df.describe().transpose # summary statistics for numerical columns

count          mean           std      min  \
person_id                    100000.0  50000.500000  28867.657797     1.00   
age                          100000.0     47.521500     15.988752     0.00   
income                       100000.0  49873.905000  46800.214127  1100.00   
household_size               100000.0      2.430900      1.075126     1.00   
dependents                   100000.0      0.898380      0.950654     0.00   
bmi                          100000.0     26.990512      4.994883    12.00   
visits_last_year             100000.0      1.927650      1.737730     0.00   
hospitalizations_last_3yrs   100000.0      0.093640      0.304848     0.00   
days_hospitalized_last_3yrs  100000.0      0.373350      1.373011     0.00   
medication_count             100000.0      1.236320      1.209358     0.00   
systolic_bp                  100000.0    117.808970     15.369187    61.00   
diastolic_bp                 100000.0     73.604770      8.900924    40.00   
ldl                          100000.0    119.975065     30.262086    30.00   
hba1c                        100000.0      5.605968      0.845996     3.54   
deductible                   100000.0   1226.725000   1019.619375   500.00   
copay                        100000.0     19.520400     10.286255    10.00   
policy_term_years            100000.0      5.517760      2.868827     1.00   
policy_changes_last_2yrs     100000.0      0.050690      0.224591     0.00   
provider_quality             100000.0      3.598920      0.594052     1.50   
risk_score                   100000.0      0.519849      0.250669     0.00   
annual_medical_cost          100000.0   3009.451907   3127.462822    55.55   
annual_premium               100000.0    582.320040    399.583722   211.67   
monthly_premium              100000.0     48.526668     33.298640    17.64   
claims_count                 100000.0      1.621780      2.029820     0.00   
avg_claim_amount             100000.0    656.512084   1072.660048     0.00   
total_claims_paid            100000.0   1377.878827   2305.464687     0.00   
chronic_count                100000.0      0.724720      0.805523     0.00   
hypertension                 100000.0      0.203450      0.402566     0.00   
diabetes                     100000.0      0.085930      0.280262     0.00   
asthma                       100000.0      0.058870      0.235382     0.00   
copd                         100000.0      0.035950      0.186166     0.00   
cardiovascular_disease       100000.0      0.051170      0.220345     0.00   
cancer_history               100000.0      0.021510      0.145078     0.00   
kidney_disease               100000.0      0.014620      0.120027     0.00   
liver_disease                100000.0      0.014770      0.120632     0.00   
arthritis                    100000.0      0.108310      0.310773     0.00   
mental_health                100000.0      0.130140      0.336459     0.00   
proc_imaging_count           100000.0      0.508530      0.749755     0.00   
proc_surgery_count           100000.0      0.158690      0.463562     0.00   
proc_physio_count            100000.0      0.508390      0.747218     0.00   
proc_consult_count           100000.0      0.509330      0.753630     0.00   
proc_lab_count               100000.0      0.509140      0.750455     0.00   
is_high_risk                 100000.0      0.367810      0.482212     0.00   
had_major_procedure          100000.0      0.169700      0.375371     0.00   

                                    25%         50%         75%         max  
person_id                    25000.7500  50000.5000  75000.2500   100000.00  
age                             37.0000     48.0000     58.0000      100.00  
income                       21100.0000  36200.0000  62200.0000  1061800.00  
household_size                   2.0000      2.0000      3.0000        9.00  
dependents                       0.0000      1.0000      1.0000        7.00  
bmi                             23.6000     27.0000  

In [7]:
df.isnull().sum()   # count missing values in each column

person_id                          0
age                                0
sex                                0
region                             0
urban_rural                        0
income                             0
education                          0
marital_status                     0
employment_status                  0
household_size                     0
dependents                         0
bmi                                0
smoker                             0
alcohol_freq                   30083
visits_last_year                   0
hospitalizations_last_3yrs         0
days_hospitalized_last_3yrs        0
medication_count                   0
systolic_bp                        0
diastolic_bp                       0
ldl                                0
hba1c                              0
plan_type                          0
network_tier                       0
deductible                         0
copay                              0
policy_term_years                  0
p

#### Our dataset has only one column with missing values:

alcohol_freq → 30,083 missing values

In [8]:
df['alcohol_freq'].value_counts(dropna=False)  # count all values, including missing ones

alcohol_freq
Occasional    45078
NaN           30083
Weekly        19833
Daily          5006
Name: count, dtype: int64

In [12]:
df['alcohol_freq'] = df['alcohol_freq'].fillna('Unknown', inplace=True)  # replace missing values with "Unknown"

C:\Users\LarTI\AppData\Local\Temp\ipykernel_15484\3414404668.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['alcohol_freq'] = df['alcohol_freq'].fillna('Unknown', inplace=True)  # replace missing values with "Unknown"


In [13]:
df.dtypes  # check the data type of each column

person_id                        int64
age                              int64
sex                             object
region                          object
urban_rural                     object
income                         float64
education                       object
marital_status                  object
employment_status               object
household_size                   int64
dependents                       int64
bmi                            float64
smoker                          object
alcohol_freq                    object
visits_last_year                 int64
hospitalizations_last_3yrs       int64
days_hospitalized_last_3yrs      int64
medication_count                 int64
systolic_bp                    float64
diastolic_bp                   float64
ldl                            float64
hba1c                          float64
plan_type                       object
network_tier                    object
deductible                       int64
copay                    

#### Identify numerical vs categorical features

In [16]:
# Numerical features
num_cols = df.select_dtypes(include=['int64', 'float64']).columns

# Categorical features
cat_cols = df.select_dtypes(include=['object']).columns

print("Numerical columns:", num_cols)
print()
print("Categorical columns:", cat_cols)


Numerical columns: Index(['person_id', 'age', 'income', 'household_size', 'dependents', 'bmi',
       'visits_last_year', 'hospitalizations_last_3yrs',
       'days_hospitalized_last_3yrs', 'medication_count', 'systolic_bp',
       'diastolic_bp', 'ldl', 'hba1c', 'deductible', 'copay',
       'policy_term_years', 'policy_changes_last_2yrs', 'provider_quality',
       'risk_score', 'annual_medical_cost', 'annual_premium',
       'monthly_premium', 'claims_count', 'avg_claim_amount',
       'total_claims_paid', 'chronic_count', 'hypertension', 'diabetes',
       'asthma', 'copd', 'cardiovascular_disease', 'cancer_history',
       'kidney_disease', 'liver_disease', 'arthritis', 'mental_health',
       'proc_imaging_count', 'proc_surgery_count', 'proc_physio_count',
       'proc_consult_count', 'proc_lab_count', 'is_high_risk',
       'had_major_procedure'],
      dtype='object')

Categorical columns: Index(['sex', 'region', 'urban_rural', 'education', 'marital_status',
       'employment_